text-serializer-gson/src/main/java/net/kyori/adventure/text/serializer/gson/SerializerFactory.java

```diff
@@ -0,0 +1,91 @@
+import net.kyori.adventure.text.format.TextDecoration;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+final class SerializerFactory implements TypeAdapterFactory {
+  static final Class<Key> KEY_TYPE = Key.class;
+  static final Class<Component> COMPONENT_TYPE = Component.class;
+  static final Class<Style> STYLE_TYPE = Style.class;
+  static final Class<ClickEvent.Action> CLICK_ACTION_TYPE = ClickEvent.Action.class;
+  static final Class<HoverEvent.Action> HOVER_ACTION_TYPE = HoverEvent.Action.class;
+  static final Class<HoverEvent.ShowItem> SHOW_ITEM_TYPE = HoverEvent.ShowItem.class;
+  static final Class<HoverEvent.ShowEntity> SHOW_ENTITY_TYPE = HoverEvent.ShowEntity.class;
+  static final Class<TextColorWrapper> COLOR_WRAPPER_TYPE = TextColorWrapper.class;
+  static final Class<TextColor> COLOR_TYPE = TextColor.class;
+  static final Class<TextDecoration> TEXT_DECORATION_TYPE = TextDecoration.class;
+  static final Class<BlockNBTComponent.Pos> BLOCK_NBT_POS_TYPE = BlockNBTComponent.Pos.class;
+
+  private final boolean downsampleColors;
+  private final LegacyHoverEventSerializer legacyHoverSerializer;
+  private final boolean emitLegacyHover;
+
+  SerializerFactory(final boolean downsampleColors, final @Nullable LegacyHoverEventSerializer legacyHoverSerializer, final boolean emitLegacyHover) {
+    this.downsampleColors = downsampleColors;
+    this.legacyHoverSerializer = legacyHoverSerializer;
+    this.emitLegacyHover = emitLegacyHover;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public <T> TypeAdapter<T> create(final Gson gson, final TypeToken<T> type) {
+    final Class<? super T> rawType = type.getRawType();
+    if(KEY_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) KeySerializer.INSTANCE;
+    } else if(STYLE_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) new StyleSerializer(this.legacyHoverSerializer, this.emitLegacyHover, gson::getAdapter);
+    } else if(CLICK_ACTION_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) IndexedSerializer.of("click action", ClickEvent.Action.NAMES);
+    } else if(HOVER_ACTION_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) IndexedSerializer.of("hover action", HoverEvent.Action.NAMES);
+    } else if(SHOW_ITEM_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) ShowItemSerializer.INSTANCE;
+    } else if(SHOW_ENTITY_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) new ShowEntitySerializer(gson::getAdapter);
+    } else if(COLOR_WRAPPER_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) TextColorWrapper.Serializer.INSTANCE;
+    } else if(COLOR_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) (this.downsampleColors ? TextColorSerializer.DOWNSAMPLE_COLOR : TextColorSerializer.INSTANCE);
+    } else if(TEXT_DECORATION_TYPE.isAssignableFrom(rawType)) {
+      return (TypeAdapter<T>) IndexedSerializer.of("text decoration", TextDecoration.NAMES);
```

Comment:
Should be returning an `INSTANCE` here too.

Reply:
I don't quite like the idea of the instances being in the same `IndexedSerializer` but I believe it's where they fit best, thoughts?
https://github.com/KyoriPowered/adventure/pull/373/commits/8e51ffd22b3c9597b51940a87e42d98c8cc1987f

Reply:
I'd just make a class with a single field for each of these, maybe? `TextDecorationSerializer`, etc